### ISSUE TITLE
Meson Build System

### ISSUE THREAD

**Issue by anarazel** [anarazel/postgres#8]
- [ ] building against system ldap fails on openldap
- [ ] rpath on netbsd broken for at least perl
- [ ] AIX symbol generation needs to be fixed
- [ ] automated testing of various linux distributions, via docker containers
  - [ ] RHEL 7
  - [ ] RHEL 8
  - [ ] Fedora Rawhide
  - [ ] Suse (See #7 )
- [ ]  automated testing via VMs
  - [ ] OpenBSD ( https://github.com/anarazel/pg-vm-images/issues/7 )
  - [ ] NetBSD ( https://github.com/anarazel/pg-vm-images/issues/6 )

**Comment by samay-sharma** [comment]
@anarazel : Can you update the Task list with this issue as well : https://github.com/mesonbuild/meson/issues/9894

**Comment by anarazel** [comment]
> @anarazel : Can you update the Task list with this issue as well : [mesonbuild/meson#9894](https://github.com/mesonbuild/meson/issues/9894) . I opened this against meson for the last meson issue.

Done. You probably now would also have the permissions to do so.

**Comment by tristan957** [comment]
Could you expand on this?

> Header install with install_subdir() is error prone (think tempfiles)

I have submitted https://github.com/mesonbuild/meson/pull/11849 fyi.

**Comment by eli-schwartz** [comment]
install_subdir() will install an entire directory tree, even if not all the files in that tree are supposed to be installed. This can also include text editor backup files etc.

https://github.com/anarazel/postgres/blob/d1491d36a3cb6ca14ff55b5d1506ba35a1e7f35a/src/include/meson.build#L163-L172

Meson 0.63.0 adds `preserve_path: true` for install_headers() which allows spelling out every single file and then installing an entire structured tree. This may or may not be a more palatable approach -- either way it's not for today as postgres still targets systems with 0.56.0, though of course any new features would have the same time delay...

**Comment by anarazel** [comment]
@tristan957 Does @eli-schwartz 's explanation answer your question?

> Meson 0.63.0 adds preserve_path: true for install_headers() which allows spelling out every single file and then installing an entire structured tree. This may or may not be a more palatable approach -- either way it's not for today as postgres still targets systems with 0.56.0, though of course any new features would have the same time delay...

It's not really a palatable approach unfortunately.  We're talking ~750 files here. Maintaining that by hand is too errorprone.

I suspect that the best way for now is a small helper script that lists the files, separates them by directory using the fs module, and than calls install_headers() for each of those directories :(.

**Comment by tristan957** [comment]
@anarazel it does. Could you explain what would be a palatable approach? Are you looking for like an `exclude_patterns` kwarg?

**Comment by anarazel** [comment]
> @anarazel it does. Could you explain what would be a palatable approach? Are you looking for like an exclude_patterns kwarg?

I am thinking of the opposite. I would like to explicitly specify *include* patterns, with everything not matching not getting installed. IME it's quite rare that one needs to install all but a few patterns, but very common to install just one type of file.

**Comment by tristan957** [comment]
Are you thinking that glob syntax would be enough?

**Comment by tristan957** [comment]
Perhaps a `patterns` kwarg and then doing the equivalent of what github does would be enough.

https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushpull_requestpull_request_targetpathspaths-ignore

The you get negated globs too

**Comment by anarazel** [comment]
> Are you thinking that glob syntax would be enough?

Depends on the complexity of the globs. I think a full regex would be good - without a regex it's hard to exclude temp files. I'd basically want to say something like include_patterns = ['^[a-zA-Z0-9]+.*\.h'], so things like ".#filename.h" etc aren't "included".

**Comment by tristan957** [comment]
Hmm. A regex module would be interesting. I think what would be annoying is that flavors of regex differ in what they support, so if Meson exposed regex through the `re` module and Muon exposed it through `libpcre2`, you would get two different impls.

Another option would be to expose https://docs.python.org/3/library/glob.html.

**Comment by tristan957** [comment]
Maybe Meson could just define no positive/negative lookaheads/behinds. That would probably be pretty simple at the end of the day.

**Comment by eli-schwartz** [comment]
Yeah there's a few reasons why I'm not a fan of regex here. Layering an exclude pattern on top of an include pattern would probably work better here though.

**Comment by tristan957** [comment]
Eli, so are you saying that we need include_directories and include_files alongside what already exists for exclude_directories and exclude_files? Or are you advocating for patterns? If so, what pattern syntax?

**Comment by anarazel** [comment]
> One possible alternative here is to have the script generate these lists, and write them to a file in the source tree e.g. headerlist.txt

I was thinking of using run_command()  and using the output to trigger install_headers() - does going via a file have advantages? I'd assume the time overhead of running that script to be neglegible compared to the other costs of running configure.

> and then use the fs.read() function to read out that list and dump it into the install_headers() function.

For 0.56 compatibility we'd need to split into multiple install_headers() invocations, one for each directory, right?

I don't love the idea of checking in something like headerslist.txt, that opens up an unnecessary avenue of merge conflicts...

**Comment by anarazel** [comment]
> Hmm. A regex module would be interesting. I think what would be annoying is that flavors of regex differ in what they support, so if Meson exposed regex through the `re` module and Muon exposed it through `libpcre2`, you would get two different impls.
> 
> Another option would be to expose https://docs.python.org/3/library/glob.html.

I wonder if a good solution could be to have a install_* option to not install hidden files. I'd assume it's very rare that people actually want to do so. And it'd largely remove the need for a full blown regex, I think.

**Comment by tristan957** [comment]
> > Hmm. A regex module would be interesting. I think what would be annoying is that flavors of regex differ in what they support, so if Meson exposed regex through the `re` module and Muon exposed it through `libpcre2`, you would get two different impls.
> > Another option would be to expose https://docs.python.org/3/library/glob.html.
> 
> I wonder if a good solution could be to have a install_* option to not install hidden files. I'd assume it's very rare that people actually want to do so. And it'd largely remove the need for a full blown regex, I think.

I wonder if `install_subdir()` should exclude meson.build files by default. With the hidden file thing, that would seem to be everything Postgres needs. But maybe there is a workflow where installing meson.build files is appropriate. Perhaps exposing that as an option would be a good idea.

**Comment by eli-schwartz** [comment]
There's no particular difference between reading it in via a run_command and doing so via fs.read(), except that in the latter case it's more idealistically pure because when you `git pull` and new files are added, the filelist.txt gets updated at the same time and this triggers a reconfigure. Doing this via run_command is, really, just the meson FAQ entry on support for globs in source lists. You can do it if you want and are willing to accept the unreliability, but meson won't officially support it.

(install_subdir() is the explicit exception to the globbing rule, as it only calculates a directory name at configure time and processes the implicit wildcard glob at install time as part of the single install rule.)

**Comment by tristan957** [comment]
Where is the Postgres workaround for https://github.com/mesonbuild/meson/issues/9493?

Given that https://github.com/mesonbuild/meson/issues/8511 is now exposed as `find_program('meson')`, I am wondering if we should add a `meson.implementation()` method that would tell us what the invoking Meson implementation is. Meson would return `'meson'`. Muon would return `'muon'`. `Meson++` would return `'meson++'`.

Otherwise, we need a better way to do https://github.com/postgres/postgres/blob/master/meson.build#L3024-L3027.

**Comment by tristan957** [comment]
I submitted https://github.com/mesonbuild/meson/pull/12014, which should be a pretty nice cleanup when the minimum version bump occurs.

**Comment by tristan957** [comment]
https://github.com/mesonbuild/meson/pull/12051 is now closed. That leaves us with two more issues to tackle for better Meson support in Postgres. The framework include issue, which I will be working on and better support for `install_subdir()` somehow...

**Comment by anarazel** [comment]
Very nice!
